src/poetry/core/version/markers.py

```diff
@@ -244,8 +247,25 @@ def validate(self, environment: dict[str, Any] | None) -> bool:
             return True
 
         # The type of constraint returned by the parser matches our constraint: either
-        # both are BaseConstraint or both are VersionConstraint.  But it's hard for mypy
-        # to know that.
+        # both are BaseConstraint or both are VersionConstraint. But it's hard for mypy
+        # to know that. That's why we need some "type ignore[arg-type]" below.
+
+        # "extra" is special because it can have multiple values at the same time.
+        # "extra == 'a'" will be true if "a" is one of the active extras.
+        # "extra != 'a'" will be true if "a" is not one of the active extras.
+        # Further, extra names are normalized for comparison.
+        if self._name == "extra":
+            extras = environment["extra"]
+            if isinstance(extras, str):
+                extras = (extras,)
+            assert isinstance(extras, Collection)
+            assert isinstance(self._constraint, Constraint)
+            normalized_value = canonicalize_name(self._constraint.value)
```

Comment:
are the environment extras guaranteed normalized at this point?

Reply:
They are normalized in the only usage so far (in the companion PR) but it might make sense to be more robust and normalize (again) since environment is just defined as `dict[str, Any]`. What do you think?

Reply:
I don't very much mind either way, so long as you're confident that there is no bug.

As a fan of having the typechecking catch as many bugs as possible I'm tempted to encourage defining a `TypedDict` or some other actually structured object to describe the marker environment: and letting the effects of that work their way through the code...  

But I am also sympathetic to the line of argument that says that if you wanted to write code that way then you didn't want to be writing python at all: and you have carefully reviewed the one place that puts `extras` here and it's fine.